src/metabase/actions/actions.clj

```diff
@@ -199,6 +202,67 @@
         (driver/with-driver driver
           (perform-action!* driver action db arg-map))))))
 
+(defn- promote-keys
+  "Promote relevant keys nested within the \"payload\" of an action lifecycle event to the top level. Why? ..."
+  [nested-map top-level-map]
+  (merge top-level-map (u/snake-keys (select-keys nested-map [:table-id]))))
+
+(defn- publish-action-invocation! [invocation-id user-id action-kw args-map]
+  (->> {:action        action-kw
+        :invocation_id invocation-id
+        :actor_id      user-id
+        :args          args-map}
+       (promote-keys args-map)
+       (events/publish-event! :event/action.invoked)))
+
+(defn- publish-action-success*! [invocation-id user-id action-kw result]
+  (->> {:action        action-kw
+        :invocation_id invocation-id
+        :actor_id      user-id
+        :result        result}
+       (promote-keys result)
+       (events/publish-event! :event/action.success)))
+
+(defn- publish-action-failure*! [invocation-id user-id action-kw msg info]
+  (->> {:action        action-kw
+        :invocation_id invocation-id
+        :actor_id      user-id
+        :error         (:error info)
+        :message       msg
+        :info          info}
+       (promote-keys info)
+       (events/publish-event! :event/action.failure)))
+
+(defn perform-with-system-events!
+  "Eventually, all calls to perform-action! should go through this... Proceeding with caution."
+  [action-kw args-map & {:as opts}]
+  (let [invocation-id (nano-id/nano-id)
+        user-id       api/*current-user-id*]
+    (publish-action-invocation! invocation-id user-id action-kw args-map)
+    (try
+      (let [result (perform-action! action-kw args-map opts)]
+        (publish-action-success*! invocation-id user-id action-kw result)
+        result)
+      (catch Exception e
+        (let [msg  (ex-message e)
+              info (ex-data e)
+              info (with-meta info (merge (meta info) {:exception e}))]
+          (publish-action-failure*! invocation-id user-id action-kw msg info)
+          (throw e))))))
+
+(defmulti action-filter-keys {:arglists '([action-kw])} identity)
+
+(defmethod action-filter-keys :default [_])
```

Comment:
this assume that the action-kw from payload matches the a column from `notification_subscription`, why make that a constraint? what about having this returns a honey sql clause/